File: Frontend/src/services/Pedido.service.js
import httpClient from "@/http-common";

// Crear un nuevo pedido
const createPedido = (pedido, token) => {
  return httpClient.post('/pedido/create', pedido, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Obtener todos los pedidos
const getAllPedidos = (token) => {
  return httpClient.get('/pedido/getAll', {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Obtener un pedido por ID
const getPedidoById = (id, token) => {
  return httpClient.get(`/pedido/getById/${id}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Actualizar un pedido por ID
const updatePedido = (id, pedido, token) => {
  return httpClient.put(`/pedido/update/${id}`, pedido, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Eliminar un pedido por ID
const deletePedido = (id, token) => {
  return httpClient.delete(`/pedido/delete/${id}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Registrar un pedido
const registrarPedido = (pedido, token) => {
  return httpClient.post('/pedido/registrar', pedido, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Confirmar un pedido y descontar el stock
const confirmarPedidoYDescontarStock = (id, token) => {
  return httpClient.put(`/pedido/confirmarPedidoYDescontarStock/${id}`, null, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Cambiar el estado de un pedido
const cambiarEstadoPedido = (id, nuevoEstado, token) => {
  return httpClient.put(`/pedido/cambiarEstadoPedido/${id}`, { nuevoEstado }, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

const obtenerPedidosCliente = (idCliente, token) => {
  return httpClient.get(`/pedido/cliente/${idCliente}`, {
    headers: {
      Authorization: `Bearer ${token}`, // Usamos el token para la autenticación
    },
  });
};


//Obtener pedidos más cercanos a una empresa
const getPedidosMasCercanosEmpresa = (idEmpresa, token, limite = 5) => {
  return httpClient.get(`/pedido/masCercanosEmpresa/${idEmpresa}?limite=${limite}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

//Obtener pedidos más lejanos por empresa
const getPedidosMasLejanosPorEmpresa = (token) => {
  return httpClient.get("/pedido/masLejanosPorEmpresa", {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Obtener pedidos que cruzan más de dos zonas de reparto
const getPedidosConMasDeDosZonas = (token) => {
  return httpClient.get("/pedido/pedidosConMasDeDosZonas", {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};


export default {
  createPedido,
  getAllPedidos,
  getPedidoById,
  updatePedido,
  deletePedido,
  registrarPedido,
  confirmarPedidoYDescontarStock,
  cambiarEstadoPedido,
  obtenerPedidosCliente,
  getPedidosMasCercanosEmpresa,
  getPedidosMasLejanosPorEmpresa,
  getPedidosConMasDeDosZonas
};

File: Backend/demo/src/main/java/com/example/demo/Service/PedidoService.java
package com.example.demo.Service;

import com.example.demo.DTO.*;
import com.example.demo.Entity.Pedido;
import com.example.demo.Repository.PedidoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;

@Service
public class PedidoService {

    @Autowired
    private PedidoRepository pedidoRepository;

    public Pedido crearPedido(Pedido pedido) {
        return pedidoRepository.crear(pedido);
    }

    //servicio getAll pedidos
    public List<PedidoDTO> obtenerTodos(int page, int size) {
        return pedidoRepository.getAll(page, size);
    }

    public PedidoDTO getById(Integer id) {
        return pedidoRepository.getById(id);
    }

    public String actualizarPedido(Pedido pedido, Integer id) {
        return pedidoRepository.update(pedido, id);
    }

    public void eliminarPedido(Integer id) {
        pedidoRepository.delete(id);
    }

    public List<ResumenPedidoDTO> obtenerResumenPorCliente(Integer idCliente){
        return pedidoRepository.obtenerResumenPorCliente(idCliente);
    }

    //--------------- PROCEDIMIENTOS---------------
    public void registrarPedido(PedidoCompletoDTO pedido) {
        pedidoRepository.registrarPedidoCompleto(pedido);
    }

    public void confirmarPedidoYDescontarStock(int id){
        pedidoRepository.confirmarPedidoYDescontarStock(id);
    }

    public void cambiarEstadoPedido(int id, String nuevoEstado){
        pedidoRepository.cambiarEstadoPedido(id, nuevoEstado);
    }

    // Método para obtener los pedidos de un cliente
    public List<PedidoDTO> obtenerPedidosPorCliente(Integer idCliente, int page, int size) {
        return pedidoRepository.getPedidosByCliente(idCliente, page, size);
    }

    public List<Pedido> getPedidosMasCercanos(Integer idEmpresa, int limite) {
        return pedidoRepository.getPedidosMasCercanos(idEmpresa, limite);
    }

    //-------------------------- CONSULTAS LAB 2 -------------------------------
    //1) Encontrar los 5 puntos de entrega más cercanos a una empresa asociada.
    public List<PedidoCercanoDTO> getPedidosMasCercanosEmpresa(Integer idEmpresa, int limite){
        return pedidoRepository.getPedidosMasCercanosEmpresa(idEmpresa, limite);
    }

    //4) Identificar el punto de entrega más lejano desde cada empresa asociada.
    public List<PedidoCercanoDTO> getPedidosMasLejanosPorEmpresa() {
        return pedidoRepository.getPedidosMasLejanosPorEmpresa();
    }

    //5) Listar todos los pedidos cuya ruta estimada cruce más de 2 zonas de reparto.
    public List<PedidoRutaDTO> getPedidosConMasDeDosZonas(){
        return pedidoRepository.getPedidosConMasDeDosZonas();
    }


}
